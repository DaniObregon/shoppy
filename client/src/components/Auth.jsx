import axios from "axios";
import React from "react";
import { signInWithPopup, GoogleAuthProvider, signOut } from "firebase/auth";
import { SignUpButton } from "./SignUpButton";
import { SignOutButton } from "./SignOutButton";
import { useNavigate } from "react-router-dom";
import { auth } from "../config/firebaseConfig";
import { useDispatch, useSelector } from "react-redux";
import { login, logout } from "../redux/slices/userSlice";
import { Flex, Text, Stack } from "@chakra-ui/react";

export const Auth = ({ onAuthSuccess }) => {
  const provider = new GoogleAuthProvider();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  // Obtener el estado del usuario desde Redux (obtiene los datos)
  const userInfo = useSelector((state) => state.userInfo);

  const signUpWithGoogle = async () => {
    try {
      const result = await signInWithPopup(auth, provider);
      const user = result.user;
      const token = await user.getIdToken();
      const userData = {
        name: user.displayName,
        email: user.email,
        role_id: 1, // Rol por defecto (client)
      };

      await axios.post("/api/auth/signup", JSON.stringify(userData), {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`, // El token se envia en un header
        },
      });

      // Actualizar el estado global de Redux con los datos del usuario
      dispatch(
        login({
          name: user.displayName,
          email: user.email,
        })
      );

      // Si la autenticación y la creación del usuario son exitosas
      if (onAuthSuccess) {
        onAuthSuccess(user); // Llamar a la función si está definida
      }
      navigate("/"); // Redirige al home o a la ruta que prefieras después de autenticarse
    } catch (error) {
      console.error("Error durante el inicio de sesión:", error);
    }
  };

  const signOutUser = async () => {
    try {
      await signOut(auth);
      dispatch(logout()); // Despachar la acción de logout en Redux
      navigate("/");
    } catch (error) {
      console.error("Error al cerrar sesión:", error);
    }
  };

  return (
    <Flex align="center" justify="center" p={4}>
      {userInfo.name ? (
        <Stack direction="row" align="center" spacing={4}>
          <Text
            p={2}
            color="gray.600"
            whiteSpace="nowrap"
            maxWidth="200px"
            overflow="hidden"
            textOverflow="ellipsis"
          >
            Hola, {userInfo.name.split(" ")[0]}
          </Text>
          <SignOutButton onClick={signOutUser} />
        </Stack>
      ) : (
        <SignUpButton onClick={signUpWithGoogle} />
      )}
    </Flex>
  );
};
