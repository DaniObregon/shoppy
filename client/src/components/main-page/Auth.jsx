import api from "../../config/axiosConfig"
import React from "react";
import { signInWithPopup, GoogleAuthProvider, signOut } from "firebase/auth";
import { SignUpButton } from "./SignUpButton";
import { SignOutButton } from "./SignOutButton";
import { useNavigate } from "react-router-dom";
import { auth } from "../../config/firebaseConfig";
import { useDispatch, useSelector } from "react-redux";
import { login, logout } from "../../redux/slices/userSlice";
import { Flex, Text, Stack } from "@chakra-ui/react";

export const Auth = ({ onAuthSuccess }) => {
  const provider = new GoogleAuthProvider();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  // Obtiene el estado del usuario desde Redux (obtiene los datos)
  const userInfo = useSelector((state) => state.userInfo);

  const signUpWithGoogle = async () => {
    try {
      const result = await signInWithPopup(auth, provider);
      const user = result.user;
      const token = await user.getIdToken();
      const userData = {
        name: user.displayName,
        email: user.email,
        role_id: 1, // Rol por defecto (client)
        token: token, // Guardamos el token aquí
      };

      //console.log("Datos que se envían al backend:", userData);

      const response = await api.post(
        "/api/auth/signup",
        JSON.stringify(userData),
        {
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`, // El token se envia en un header
          },
        }
      );

      //console.log("Respuesta del backend:", response.data);

      // Accede a las propiedades del objeto `user` desde la respuesta del backend
      const { name, email, role } = response.data.user;

      //console.log("Datos recibidos del backend:", { name: name, email: email, role: role });

      // Actualizar el estado global de Redux con los datos del usuario incluyendo el token
      dispatch(
        login({
          name: name,
          email: email,
          role_id: role,
          token: token,
        })
      );

      // Si la autenticación y la creación del usuario son exitosas
      if (onAuthSuccess) {
        onAuthSuccess(user); // Llamar a la función si está definida
      }
      navigate("/"); // Redirige al home después de autenticarse (o cambiar ruta)
    } catch (error) {
      console.error("Error durante el inicio de sesión:", error);
    }
  };

  const signOutUser = async () => {
    try {
      await signOut(auth);
      dispatch(logout()); // Despachar la acción de logout en Redux
      navigate("/");
    } catch (error) {
      console.error("Error al cerrar sesión:", error);
    }
  };

  return (
    <Flex align="center" justify="center" p={4}>
      {userInfo.name ? (
        <Stack direction="row" align="center" spacing={4}>
          <Text
            p={2}
            color="gray.600"
            whiteSpace="nowrap"
            maxWidth="200px"
            overflow="hidden"
            textOverflow="ellipsis"
            fontSize={"m"}
          >
            Hola, {userInfo.name.split(" ")[0]}
          </Text>
          <SignOutButton onClick={signOutUser} />
        </Stack>
      ) : (
        <SignUpButton onClick={signUpWithGoogle} />
      )}
    </Flex>
  );
};
