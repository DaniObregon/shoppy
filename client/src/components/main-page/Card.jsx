import React from "react";
import { useNavigate } from "react-router-dom";
import { chakra, Box, Flex } from "@chakra-ui/react";
import { AddToCartButton } from "./AddToCartButton";

export const Card = ({ id, brand, model, description, price, imgUrl }) => {
  const navigate = useNavigate();

  const handleNavigate = () => {
    navigate(`/product/${id}`);
  };

  return (
    <Flex
      direction="column"
      justifyContent="center"
      alignItems="center"
      w="full"
      mx="auto"
      mb={4}
      transition="transform 0.15s ease-in-out"
      _hover={{
        transform: "scale(1.015)",
        cursor: "pointer",
      }}
      onClick={handleNavigate} // Redirige cuando se hace clic en la card
    >
      <Box
        bg="gray.300"
        h={{ base: 56, md: 64 }} //Imagen m치s grande para mantener la proporci칩n
        w="full"
        rounded="lg"
        shadow="md"
        bgSize="cover"
        bgPos="center"
        style={{
          backgroundImage: `url(${imgUrl})`,
        }}
      ></Box>

      <Box
        w={{ base: "80%", md: "70%" }} // Caja de datos peque침a y centrada
        bg="white"
        _dark={{ bg: "gray.800" }}
        mt={-10} // Superpone la caja de datos en la imagen
        shadow="lg"
        rounded="lg"
        overflow="hidden"
      >
        <chakra.h3
          py={2}
          textAlign="center"
          fontWeight="bold"
          textTransform="uppercase"
          color="gray.800"
          _dark={{ color: "white" }}
          letterSpacing={1}
        >
          {brand}
        </chakra.h3>
        <chakra.p
          py={1}
          textAlign="center"
          fontWeight="medium"
          color="gray.600"
          _dark={{ color: "gray.300" }}
          fontSize="sm"
        >
          {model}
        </chakra.p>
        <chakra.p
          py={1}
          textAlign="center"
          fontWeight="medium"
          color="gray.600"
          _dark={{ color: "gray.300" }}
          fontSize="sm"
          sx={{
            overflow: "hidden",
            display: "-webkit-box",
            WebkitLineClamp: "3",
            WebkitBoxOrient: "vertical",
            wordBreak: "break-word", // Corta palabras largas autom치ticamente
            overflowWrap: "break-word", // Compatibilidad adicional
          }}
        >
          {description}
        </chakra.p>
        <Flex
          alignItems="center"
          justifyContent="space-between"
          py={2}
          px={3}
          bg="gray.200"
          _dark={{ bg: "gray.700" }}
        >
          <chakra.span
            fontWeight="bold"
            color="gray.800"
            _dark={{ color: "gray.200" }}
          >
            ${price}
          </chakra.span>
          <AddToCartButton id={id} />
        </Flex>
      </Box>
    </Flex>
  );
};
