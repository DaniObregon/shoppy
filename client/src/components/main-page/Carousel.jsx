import React, { useEffect, useState } from "react";
import { Box, Flex, Image, Text } from "@chakra-ui/react";
import { slides } from "../../data/carouselSlides";

export const Carousel = () => {
  const [currentSlide, setCurrentSlide] = useState(0);
  const [visibleSlides, setVisibleSlides] = useState(slides);
  const [isPaused, setIsPaused] = useState(false); // Estado para manejar si la transición está pausada
  const [isHovered, setIsHovered] = useState(false); // Estado para manejar si el mouse está sobre el carrusel

  const slidesCount = visibleSlides.length;
  const carouselStyle = {
    transition: "all 1.5s",
    ml: `-${currentSlide * 100}%`,
  };

  const SLIDES_INTERVAL_TIME = 4000;
  const ANIMATION_DIRECTION = "right";

  useEffect(() => {
    const prevSlide = () => {
      setCurrentSlide((s) => (s === 0 ? slidesCount - 1 : s - 1));
    };

    const nextSlide = () => {
      setCurrentSlide((s) => (s === slidesCount - 1 ? 0 : s + 1));
    };

    const automatedSlide = setInterval(() => {
      if (!isPaused) {
        ANIMATION_DIRECTION.toLowerCase() === "left"
          ? prevSlide()
          : nextSlide();
      }
    }, SLIDES_INTERVAL_TIME);

    return () => clearInterval(automatedSlide);
  }, [slidesCount, isPaused]);

  const handleMouseEnter = () => {
    setIsPaused(true);
    setIsHovered(true);
  };

  const handleMouseLeave = () => {
    setIsPaused(false);
    setIsHovered(false);
  };

  // Función para manejar el error de carga de imágenes y eliminar el slide
  const handleImageError = (index) => {
    setVisibleSlides((prevSlides) =>
      prevSlides.filter((_, slideIndex) => slideIndex !== index)
    );
  };

  return (
    <Flex
      id="carousel"
      role="region"
      aria-label="carousel"
      w="full"
      bg="#edf3f8"
      _dark={{ bg: "#3e3e3e" }}
      p={0}
      alignItems="center"
      justifyContent="center"
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <Flex w="full" overflow="hidden">
        <Flex pos="relative" h="500px" w="full" {...carouselStyle}>
          {visibleSlides.map((slide, index) => (
            <Box
              key={`slide-${index}`}
              flex="none"
              boxSize="full"
              shadow="md"
              transition="transform 0.3s ease"
              style={{
                transform: isHovered ? "translateY(-5px)" : "translateY(0px)",
              }}
            >
              <Text
                color="white"
                fontSize="xs"
                p="8px 12px"
                pos="absolute"
                top="0"
                whiteSpace="nowrap"
              >
                {index + 1} / {slidesCount}
              </Text>
              <Image
                src={slide.img}
                alt={`Slide ${index + 1}`}
                boxSize="full"
                objectFit="cover"
                style={{
                  opacity: isHovered ? "0.8" : "1",
                  transition: "opacity 0.3s ease",
                }}
                onError={() => handleImageError(index)} // Maneja el error de carga de imagen eliminando el slide
              />
            </Box>
          ))}
        </Flex>
      </Flex>
    </Flex>
  );
};
