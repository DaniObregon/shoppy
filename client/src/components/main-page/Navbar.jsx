import React from "react";
import {
  Box,
  Flex,
  IconButton,
  Stack,
  Collapse,
  useColorModeValue,
  useDisclosure,
} from "@chakra-ui/react";
import { HamburgerIcon, CloseIcon } from "@chakra-ui/icons";
import { Auth } from "./Auth";
import { DarkModeToggle } from "./DarkModeToggle";
import { Logo } from "./Logo";
import { DesktopNav } from "./DesktopNav";
import { MobileNav } from "./MobileNav";

export const Navbar = () => {
  const { isOpen, onToggle } = useDisclosure();

  return (
    <Box mt="60px" marginBottom={1} marginTop={5}>
      <Flex
        bg={useColorModeValue("white", "gray.800")}
        color={useColorModeValue("gray.600", "white")}
        minH={"60px"}
        py={{ base: 2 }}
        px={{ base: 10 }}
        borderBottom={1}
        borderTop={1}
        borderStyle={"solid"}
        borderColor={useColorModeValue("gray.200", "gray.900")}
        align={"center"}
        justify={"space-between"}
        marginTop={8}
        marginBottom={1}
        paddingTop="9px"
        border={-10}
      >
        <Flex
          flex={{ base: 1, md: "auto" }}
          ml={{ base: -2 }}
          display={{ base: "flex", md: "none" }}
        >
          <IconButton
            onClick={onToggle}
            icon={
              isOpen ? <CloseIcon w={3} h={3} /> : <HamburgerIcon w={5} h={5} />
            }
            variant={"ghost"}
            aria-label={"Toggle Navigation"}
          />
        </Flex>
        <Flex flex={{ base: 1 }} justify={{ base: "center", md: "start" }}>
          <Logo />
          <Flex
            display={{ base: "none", md: "flex" }}
            ml={10}
            alignItems={"center"}
          >
            <DesktopNav />
          </Flex>
        </Flex>

        <Stack
          flex={{ base: 1, md: 0 }}
          justify={"flex-end"}
          direction={"row"}
          spacing={6}
          alignItems="center" // Asegura que todo estÃ© alineado verticalmente
        >
          <Auth /> {/* Este componente maneja tanto login como logout */}
          <DarkModeToggle />
        </Stack>
      </Flex>

      <Collapse in={isOpen} animateOpacity>
        <MobileNav />
      </Collapse>
    </Box>
  );
};

<DesktopNav />;
<MobileNav />;
