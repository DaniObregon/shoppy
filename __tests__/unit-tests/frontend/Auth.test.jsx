import "@testing-library/jest-dom";
import React from "react";
import { render, screen } from "@testing-library/react";
import { Auth } from "../../../client/src/components/main-page/Auth";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";
import { ChakraProvider } from "@chakra-ui/react";

// Mockear módulos relacionados con Firebase
jest.mock("../../../client/src/config/firebaseConfig", () => ({
  auth: require("../../../__mocks__/firebaseConfig").auth,
}));

// Crear un mock del store de Redux
const mockStore = configureStore([]);
const initialState = {
  userInfo: { name: "", email: "", role_id: null },
};

const loggedState = {
  userInfo: { name: "giuseppe", email: "giuseppe@mail.com", role_id: 2 },
};

describe("Auth component", () => {
  let store;

  // beforeEach(() => {
  //   store = mockStore(initialState);
  // });

  // -----1 - Prueba de Renderización
  test("should render the SignOutButton when the user is logged in", () => {
    store = mockStore(loggedState);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ChakraProvider>
            <Auth />
          </ChakraProvider>
        </BrowserRouter>
      </Provider>
    );
    const SignOutButton = screen.getByRole("button", { name: /salir/i });
    expect(SignOutButton).toBeInTheDocument();
  });

  // 1 - Prueba de Renderización
  test("should render the SignUpButton when the user is not logged in", () => {
    store = mockStore(initialState);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ChakraProvider>
            <Auth />
          </ChakraProvider>
        </BrowserRouter>
      </Provider>
    );
    const signUpButton = screen.getByRole("button", { name: /ingresar/i });
    expect(signUpButton).toBeInTheDocument();
  });

  // 2 - Prueba de Visibilidad en Diferentes Pantallas
  test("should be visible on mobile and larger screens", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ChakraProvider>
            <Auth />
          </ChakraProvider>
        </BrowserRouter>
      </Provider>
    );
    const signUpButton = screen.getByRole("button", { name: /ingresar/i });

    // Simular tamaño de pantalla pequeña (móvil)
    global.innerWidth = 480;
    global.dispatchEvent(new Event("resize"));
    expect(signUpButton).toBeVisible();

    // Simular tamaño de pantalla grande (desktop)
    global.innerWidth = 1024;
    global.dispatchEvent(new Event("resize"));
    expect(signUpButton).toBeVisible();
  });

  // 3 - Snapshot Testing
  test("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ChakraProvider>
            <Auth />
          </ChakraProvider>
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
