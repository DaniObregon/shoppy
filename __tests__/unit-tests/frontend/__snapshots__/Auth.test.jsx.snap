// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-177jhv5"
  >
    <button
      class="chakra-button css-1e2mvam"
      type="button"
    >
      Ingresar
    </button>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
